plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveInsertOverwrite.java

```diff
@@ -0,0 +1,216 @@
+                "bucketed_by = ARRAY['nationkey']",
+                "bucket_count = 3"));
+        assertInsertFailsOnUnsupportedWritingToTarget(
+                testTable,
+                "Overwriting unpartitioned table not supported when writing directly to target directory");
+    }
+
+    @Test
+    public void testInsertOverwritePartitionedTable()
+    {
+        String testTable = getTestTableName();
+        computeActual(getCreateTableStatement(
+                testTable,
+                "partitioned_by=ARRAY['regionkey']"));
+        copyTpchNationToTable(testTable);
+        assertOverwritePartition(testTable);
+    }
+
+    @Test
+    public void testInsertOverwritePartitionedAndBucketedTable()
+    {
+        String testTable = getTestTableName();
+        computeActual(getCreateTableStatement(
+                testTable,
+                "partitioned_by=ARRAY['regionkey']",
+                "bucketed_by = ARRAY['nationkey']",
+                "bucket_count = 3"));
+        copyTpchNationToTable(testTable);
+        assertOverwritePartition(testTable);
+    }
+
+    @Test
+    public void testInsertOverwritePartitionedAndBucketedExternalTable()
+    {
+        String testTable = getTestTableName();
+        // Store table data in data lake bucket
+        computeActual(getCreateTableStatement(
+                testTable,
+                "partitioned_by=ARRAY['regionkey']",
+                "bucketed_by = ARRAY['nationkey']",
+                "bucket_count = 3"));
+        copyTpchNationToTable(testTable);
+
+        // Map this table as external table
+        String externalTableName = testTable + "_ext";
+        computeActual(getCreateTableStatement(
+                externalTableName,
+                "partitioned_by=ARRAY['regionkey']",
+                "bucketed_by = ARRAY['nationkey']",
+                "bucket_count = 3",
+                format("external_location = 's3a://%s/%s/%s/'", this.bucketName, HIVE_TEST_SCHEMA, testTable)));
+        copyTpchNationToTable(testTable);
+        assertOverwritePartition(externalTableName);
+    }
+
+    protected void assertInsertFailsOnUnsupportedWritingToTarget(String testTable, String expectedMessageRegExp)
+    {
+        assertQueryFails(
+                format("INSERT INTO %s " +
+                                "SELECT name, comment, nationkey, regionkey " +
+                                "FROM tpch.tiny.nation",
+                        testTable),
+                expectedMessageRegExp);
+        computeActual(format("DROP TABLE %s", testTable));
+    }
+
+    protected void assertOverwritePartition(String testTable)
+    {
+        computeActual(format(
+                "INSERT INTO %s VALUES " +
+                        "('POLAND', 'Test Data', 25, 5), " +
+                        "('CZECH', 'Test Data', 26, 5)",
+                testTable));
+        MaterializedResult materializedRows = computeActual(format("SELECT count(*) FROM %s WHERE regionkey=5", testTable));
```

Comment:
nit `regionkey = 5`

Reply:
Also why do you need `MaterializedRow` here. You should be good with:
```
        assertQuery("SELECT count(*) FROM " + testTable + " WHERE regionkey=5", "VALUES CAST(2 AS BIGINT)");
```

(other places too)

Reply:
assertQuery is void. So I refactored to use query("...").assertThat